NebulaAPI/Extensions/PlayerCollectionExtensions.cs

```diff
@@ -0,0 +1,115 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using NebulaAPI.GameState;
+using NebulaAPI.Networking;
+
+namespace NebulaAPI.Extensions;
+
+public static class PlayerCollectionExtensions
+{
+    // public static IReadOnlyCollection<IPlayerData> GetAllPlayerDataIncludingHost(
+    //     this IReadOnlyCollection<INebulaPlayer> players, ILocalPlayer host)
+    // {
+    //     var all = players
+    //         .Select(p => p.Data)
+    //         .ToList();
+    //     all.Add(host.Data);
+    //
+    //     return all;
+    // }
+
+    public static IReadOnlyDictionary<INebulaConnection, INebulaPlayer> GetByConnectionStatus(
+        this IReadOnlyDictionary<INebulaConnection, INebulaPlayer> playerConnections, EConnectionStatus status)
+    {
+        return playerConnections
+            .Where(kvp => kvp.Key.ConnectionStatus == status)
+            .ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
+    }
+
+    public static IReadOnlyCollection<INebulaPlayer> GetByConnectionStatus(
+        this IReadOnlyCollection<INebulaPlayer> players, EConnectionStatus status)
+    {
+        return players
+            .Where(value => value.Connection.ConnectionStatus == status)
+            .ToList();
+    }
+
+    public static INebulaPlayer GetByPlayerId(
+        this IReadOnlyCollection<INebulaPlayer> players, ushort id)
+    {
+        return players.First(p => p.Id == id);
+    }
+
+    /// <summary>
+    /// 
+    /// </summary>
+    /// <param name="players"></param>
+    /// <param name="conn"></param>
+    /// <returns>
+    /// The player corresponding to the connection handle.
+    /// Can return null, if the player disconnects before the call.
+    /// </returns>
+    public static INebulaPlayer GetByConnectionHandle(
+        this IReadOnlyCollection<INebulaPlayer> players, INebulaConnection conn)
+    {
+        return players.FirstOrDefault(p => p.Connection.Equals(conn));
+    }
+
+    public static INebulaPlayer GetByUsername(
+        this IReadOnlyCollection<INebulaPlayer> players, string username)
+    {
+        return players.First(p => p.Data.Username == username);
+    }
```

Comment:
Here GetByConnectionHandle can return null but GetByUsername has the same behavior as GetByPlayerId where it will throw if the player is not found. So same question as before, is it better to throw here or return null? And shouldn't the behavior be the same for all of these?

Reply:
Fixed, all of them will now no longer throw